cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -430,7 +433,33 @@ template <>
 hash_value_type CUDA_DEVICE_CALLABLE
 SparkMurmurHash3_32<numeric::decimal128>::operator()(numeric::decimal128 const& key) const
 {
-  return this->compute<__int128_t>(key.value());
+  // Generates the Spark murmur3 hash value, mimicing the conversion
+  // java.math.BigDecimal.valueOf(unscaled_value, _scale).unscaledValue().toByteArray()
+  // https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L381
+  __int128_t const val    = key.value();
+  int8_t const* data      = reinterpret_cast<int8_t const*>(&val);
+  int8_t const sign_bit   = data[15] & 0x80;
+  int8_t const zero_value = sign_bit ? 0xff : 0x00;
+  int32_t length          = 15;
+
+  // Search for first non-zero byte in the unscaled value, shortening the hashed data
+  for (; length >= 0 && data[length] == zero_value; --length) {}
```

Comment:
Let's try to eliminate this raw for loop. Can we use a `std::find_if_not` in reverse here? Something like:
```c++
auto first_nonzero_byte = std::find_if_not(std::reverse_iterator(data + sizeof(numeric::decimal128)), data, [zero_value](int8_t const& v){ return v == zero_value; });
auto const length = std::distance(data, first_nonzero_byte);
```
Note that `length == 16` if all bytes are equal to `zero_value`. The returned iterator is not `const` because I update it in a later suggestion, see below.

Reply:
Also `length` is probably not needed, see subsequent suggestions.

Reply:
Resolved the `length` discussion offline and decided that it makes sense to keep given its usage. Code is more readable relative to the alternatives discussed.